lib/base/Page.ts

```diff
@@ -0,0 +1,278 @@
+    "total",
+    "uri",
+  ];
+
+  /**
+   * Get the url of the previous page of records
+   *
+   * @return {string|undefined} url of the previous page
+   */
+  getPreviousPageUrl(): string {
+    if (
+      "meta" in this._payload &&
+      "previous_page_url" in this._payload.meta &&
+      this._payload.meta.previous_page_url
+    ) {
+      // jshint ignore:line
+      return this._payload.meta.previous_page_url; // jshint ignore:line
+    }
+
+    if (
+      "previous_page_uri" in this._payload &&
+      this._payload.previous_page_uri
+    ) {
+      // jshint ignore:line
+      return this._version._domain.absoluteUrl(this._payload.previous_page_uri); // jshint ignore:line
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Get the url of the next page of records
+   *
+   * @return {string|undefined} url of the next page
+   */
+  getNextPageUrl(): string {
+    if (
+      "meta" in this._payload &&
+      "next_page_url" in this._payload.meta &&
+      this._payload.meta.next_page_url
+    ) {
+      // jshint ignore:line
+      return this._payload.meta.next_page_url; // jshint ignore:line
+    }
+
+    if ("next_page_uri" in this._payload && this._payload.next_page_uri) {
+      // jshint ignore:line
+      return this._version._domain.absoluteUrl(this._payload.next_page_uri); // jshint ignore:line
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Build a new instance given a json payload
+   * @abstract
+   *
+   * @param {object} payload - Payload response from the API
+   * @return {object} instance of a resource
+   */
+  getInstance(payload: any) {
+    throw new Error(
+      "Page.get_instance() must be implemented in the derived class"
+    );
+  }
+
+  /**
+   * Load a list of records
+   *
+   * @param  {object} resources json payload of records
+   * @return {Array} list of resources
+   */
+  loadInstances(resources: TResource[] | TInstance[]): TInstance[] {
+    let instances: TInstance[] = [];
+    Object.keys(resources).forEach((key) => {
+      instances[key] = this.getInstance(resources[key]);
+    });
+    return instances;
+  }
+
+  /**
+   * Fetch the next page of records
+   *
+   * @return {promise} promise that resolves to next page of results
+   */
+  nextPage(): Promise<Page<TVersion, TPayload, TResource, TInstance>> {
+    if (!this.nextPageUrl) {
+      return undefined;
+    }
+
+    var reqPromise = this._version._domain.twilio.request({
+      method: "get",
+      uri: this.nextPageUrl,
+    });
+
+    var nextPagePromise: Promise<
+      Page<TVersion, TPayload, TResource, TInstance>
+    > = reqPromise.then(
+      function (response) {
+        return new this.constructor(this._version, response, this._solution);
+      }.bind(this)
+    );
+
+    return nextPagePromise;
+  }
+
+  /**
+   * Fetch the previous page of records
+   *
+   * @return {promise} promise that resolves to previous page of results
+   */
+  previousPage(): Promise<Page<TVersion, TPayload, TResource, TInstance>> {
```

Comment:
Same as above: Should include undefined as return type?

Reply:
Done